feat: (platform) introduce step-input

dimamarksman opened the pull request:
#### Please provide a link to the associated issue.

https://github.com/SAP/fundamental-ngx/issues/2887

#### Please provide a brief summary of this pull request.

This is the platform `number-step-input` component created from scratch. The Core step-input component hasn't been used cause at the moment of working on it core step input was considered as unstable and required some fixes. 
It consists of base step-input class which intended to be used to create derivative step-input components such as Number, Money and UnitOfMeasure.

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [x] Documentation Examples
- [x] Stackblitz works for all examples

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/SAP/fundamental-ngx?pullRequest=3005) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/SAP/fundamental-ngx?pullRequest=3005) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/SAP/fundamental-ngx?pullRequest=3005) it.</sub>

fkolar commented on libs/platform/src/lib/components/form/step-input/base.step-input.ts:
```diff
@@ -0,0 +1,477 @@
+    @Input()
+    set max(max: number) {
+        this._max = !isNaN(max) ? max : Number.MAX_VALUE;
+        this._calculateCanIncrement();
```
In base input there is `stateChanges` and you need to emit event up to parent  (form-field) which listens for changes and detect chances once.

dimamarksman replied:
`max/min` options do not change control value nor change error state, they enable/disable ability to increment/decrement.
If so it's not clear for me why we should trigger event in the `stateChanges` subject.

fkolar commented on libs/platform/src/lib/components/form/step-input/base.step-input.ts:
```diff
@@ -0,0 +1,477 @@
+    /** Horizontally aligns value inside input */
+    @Input()
+    set align(align: StepInputAlign) {
+        this._align$.next(align);
```
The same here with the stateChanges and other more significant inputs that can change after component is initialized.

dimamarksman replied:
Why do we need to notify parent when `align` changes?
`Align` does not influence on form state, it just changes style in the template

fkolar replied:
We need to inform a parent to trigger change detector if some state change after component is created. You are right that align is going to be set only once we are not going to change it afterwards

fkolar commented on libs/platform/src/lib/components/form/step-input/base.step-input.ts:
```diff
@@ -0,0 +1,477 @@
+
+    /** @hidden */
+    detectChanges(): void {
+        this._cd.detectChanges();
```
Do you need to force to detect Changes immediately?  Instead of Mar k?

again if you check out the FormField that listens for changes, it does trigger detect/mark once for its input control

dimamarksman replied:
Removed this method

fkolar commented on libs/platform/src/lib/components/form/step-input/base.step-input.ts:
```diff
@@ -0,0 +1,477 @@
+            )
+            .subscribe((align) => {
+                this._align = align;
+                this.detectChanges();
```
again here. Let's correctly figureout when to emit event to formField, when to markForCheck and when to call detectChanges.

dimamarksman replied:
Changed it to `markForCheck()`.
We need `markForCheck` here because `rtlService` is out of the component context.

fkolar commented on apps/docs/src/app/platform/component-docs/platform-forms/step-input/platform-step-input-examples/platform-number-step-input-template-example.component.html:
```diff
@@ -0,0 +1,27 @@
+    <div class="fd-page__content fd-has-background-color-neutral-2">
+        <fdp-form-group #formGroup [multiLayout]="true">
+            <fdp-form-field
+                #formField
```
you dont probably need this handle #formField as you are not accessing anything in there.

dimamarksman replied:
removed it

dimamarksman commented:
> Another question for the examples.. When you define your own validator (required) shouldnt at some point trigger another form input state like error, maybe you can simulate that with other example to make sure it can add red border around the input.
> 
> When I delete a value it defaults itself to zero it does not have chance to get into error. maybe when you call validator `stepInputQtyValidators` you can really add some QTY validity like qty > 0 or something.

@fkolar I changed reactive form example, removed min/max options and added min/max validation to see the error.
